cdc/sink/producer/kafka/kafka.go

```diff
@@ -395,80 +400,127 @@ func (k *kafkaSaramaProducer) run(ctx context.Context) error {
-				"partition number(%d) assigned in sink-uri is more than that of topic(%d)", partitionNum, topicDetail.NumPartitions)
+		return err
+	}
+
+	var realPartitionCount int32 = 0
+	info, ok := topics[topic]
+	if ok {
+		realPartitionCount = info.NumPartitions
+	}
+
+	// if user specify cdc to create the topic, we must make sure
+	// that `partition-num` and `replication-factor` is given
+	if config.AutoCreate {
+		// if the specified topic name already exist, we should make sure that
+		// topic's `max.message.bytes` is not less than given `max-message-size`
+		if ok {
+			if a, ok := info.ConfigEntries["max.message.bytes"]; ok {
+				topicMaxMessageBytes, err := strconv.Atoi(*a)
+				if err != nil {
+					return cerror.WrapError(cerror.ErrKafkaNewSaramaProducer, err)
+				}
+				if topicMaxMessageBytes < config.MaxMessageBytes {
+					return cerror.ErrKafkaInvalidConfig.GenWithStack(
+						"topic already exist, and max.message.size(%d) less than max-message-size(%d)."+
+							"Please make sure max-message-size not greater than topic max.message.size",
+						topicMaxMessageBytes, config.MaxMessageBytes)
+				}
+			}
+
+			config.PartitionNum = realPartitionCount
+			log.Warn("topic already exist", zap.String("topic", topic))
+			return nil
 		}
-	} else {
-		if partitionNum == 0 {
-			partitionNum = defaultPartitionNum
-			log.Warn("topic not found and partition number is not specified, using default partition number", zap.String("topic", topic), zap.Int32("partition_num", partitionNum))
+		// this could happen if user does not specify the `partition-num` in the sink uri.
+		if config.PartitionNum == 0 {
+			config.PartitionNum = defaultPartitionNum
+			log.Warn("partition-num is not set, use the default partition count",
+				zap.String("topic", topic), zap.Int32("partitions", config.PartitionNum))
 		}
-		log.Info("create a topic", zap.String("topic", topic),
-			zap.Int32("partition_num", partitionNum),
-			zap.Int16("replication_factor", config.ReplicationFactor))
+		maxMessageBytes := strconv.Itoa(config.MaxMessageBytes)
 		err := admin.CreateTopic(topic, &sarama.TopicDetail{
-			NumPartitions:     partitionNum,
+			NumPartitions:     config.PartitionNum,
 			ReplicationFactor: config.ReplicationFactor,
+			ConfigEntries: map[string]*string{
+				"max.message.bytes": &maxMessageBytes,
+			},
 		}, false)
-		// TODO idenfity the cause of "Topic with this name already exists"
+		// TODO identify the cause of "Topic with this name already exists"
 		if err != nil && !strings.Contains(err.Error(), "already exists") {
-			return 0, cerror.WrapError(cerror.ErrKafkaNewSaramaProducer, err)
+			return err
 		}
+		return nil
+	}
+
+	// if `auto-create-topic` is disabled by user, we would assume topic should have already created
+	if !ok {
+		return cerror.ErrKafkaInvalidConfig.GenWithStack("auto-create-topic is false, and topic not found")
+	}
+
+	if config.PartitionNum == 0 {
+		config.PartitionNum = realPartitionCount
```

Comment:
`realPartitionCount` could be zero here?

Reply:
If the user does not specify `partition-num` in the sink uri, it will be 0, which means uninitialized.

Reply:
Well, it must be `AutoCreate == false` and topic exists when running at L479.
Can we extract several independent part, such as
- determine and check config.PartitionNum
- create topic if needed
- check max message bytes if needed